### DISCUSSION TITLE
Unions in negative position - is it possible?

### DISCUSSION THREAD

**Discussion by erszcz** [LPTK/simple-sub#80] [General]
Thanks for a great paper and a succinct reference implementation! I'm wondering about the limitation stated in section 2.4 of _The Simple Essence of Algebraic Subtyping_:

> There is an important caveat to add to the definition of types we gave above: these types cannot actually be used freely within type expressions. [...] In particular, unions are positive types (and may not appear in negative position) and intersections are negative types (and may not appear in positive position).

The text is even more explicit later:

> In fact, in MLsub, one _cannot_ express the type of a function which takes “either an integer or a string”: type int ⊔ string → 𝜏 is illegal because it has the int ⊔ string union in negative position.

I'm wondering if simple(r)-sub could be applied to [Erlang](https://www.erlang.org/doc/reference_manual/typespec.html) and [Elixir](https://hexdocs.pm/elixir/typespecs.html). These languages allow for unions in the type syntax and for a very limited form of intersections, i.e. multiple type annotation clauses for functions. An assumption we can make, if that makes things any simpler, is the presence of annotations on top-level functions. Would it be possible to lift the polarity restriction on unions (and just unions, not intersections)? What would be the cost of doing it if it's possible at all?

**Comment by LPTK** [comment]
Hey Radek,

Thanks for your interest in this work!

As a matter of fact, I've already generalized Simple-sub to supporting first-class unions and intersections. There's a paper that just came out about it: [MLstruct: Principal Type Inference in a Boolean Algebra of Structural Types](https://lptk.github.io/files/[v6.2]%20mlstruct.pdf).

The corresponding project is _MLstruct_ – try it out here: https://hkust-taco.github.io/mlstruct/

The main downside is that this is much harder (as in, an order of magnitude harder) to implement than Simple-sub, especially the parts about normal forms and type simplification. But it _does_ retain the principal type inference property!

One limitation is that we can't have "smart" record unions like in TypeScript: in MLstruct `{ tag: 0; value: Str } | { tag: 1; value: Int }` and `{ tag: 0 | 1; value: Str | Int }` are equivalent.

Another limitation is that while intersections can be used for mixins and field refinements, they cannot be used for function overloading as in TypeScript and what I think you're referring to in "multiple type annotation clauses for functions".

Hope this helps!

**Reply by erszcz** [reply]
Wow, that's fantastic!

I don't think "smart" record unions would be a problem, but function overloading is exactly what I had in mind. Code like this is quite common in Erlang / Elixir - the two clauses of `spec` delimited by `;` are what I meant by "limited form of intersections" - apart from that, intersection types are not allowed:

```erlang
-spec f(bar) -> bar;
       (list()) -> baz | qux.
f(bar) -> bar;
f([]) -> baz;
f([_|_]) -> qux.
```

I'll read the MLstruct paper and see if we could adapt this approach. Thanks for all the pointers!

**Reply by LPTK** [reply]
Yeah, the problem with this is that it leads to ambiguous constraints. There is no good way of locally solving `(S -> T) & (U -> V) <: A -> B` – in general, you'd need to either explore both branches and make an early decision (which completely destroys the flow of principal type inference) or you'd need to do backtracking, i.e., try each possibility as well as everything that follows it and go back to this decision if it didn't work out. The latter is arguably even worse because it doesn't scale at all (bigger expressions may take exponentially longer to type check) and it returns a set of possible types instead of a single type.

I've had some ideas for MLscript to strike a good middle-ground, but I haven't had time to try them out yet!
